### PR TITLE
Add read attribute positive and negative unit tests

### PR DESCRIPTION
**Pull request by yunhanw-google**
#### Problem
Need attribute test for Read

#### Change overview
--Add read attribute positive test
--Add read attribute negative test and fix the exchange context close
--Add more test for clusterInfo to check if A include B.
--Minor cleanup for awaiting response and onreadrequest rename
#### Testing
--Add read attribute positive test to read the right attribute path
--Add read attribute negative test to read the incorrect attribute path
--Add more test for clusterInfo to check if A include B.

### CODE REVIEW

File: src/app/ReadHandler.cpp [hunk 1]
```diff
@@ -51,8 +51,17 @@ CHIP_ERROR ReadHandler::Init(InteractionModelDelegate * apDelegate, Messaging::E
     return err;
 }
 
-void ReadHandler::Shutdown()
+void ReadHandler::Shutdown(bool aAbort)
```
Comment (andy31415) [inline src/app/ReadHandler.cpp:54]: any chance to have a signature using some 'flags enum' or similar instead of a bool. Without reading the signature, a code that reads:

```
foo->Shutdown(true);
bar->Shutdown(false);
```

is soewhat harder to read than something like:

```
foo->Shutdown(ReadHandler::kAbortCurrentExchange);
bar->Shutdown(ReadHandler::kKeepCurrentExchange);
```

File: src/app/ReadHandler.cpp [hunk 2]
```diff
@@ -103,22 +111,11 @@ CHIP_ERROR ReadHandler::OnStatusReport(Messaging::ExchangeContext * apExchangeCo
...
-        MoveToState(HandlerState::Reporting);
-    }
-
-exit:
-    ChipLogFunctError(err);
-    if (err != CHIP_NO_ERROR)
-    {
-        Shutdown();
-    }
     return err;
```
Comment (andy31415) [inline src/app/ReadHandler.cpp:119]: how about just:

```cpp
return mpExchangeCtx->SendMessage(Protocols::InteractionModel::MsgType::ReportData, std::move(aPayload),
￼                                 Messaging::SendFlags(Messaging::SendMessageFlags::kExpectResponse));
```

and remove the err local variable.

File: src/app/ReadHandler.h [hunk 3]
```diff
@@ -68,7 +68,7 @@ class ReadHandler : public Messaging::ExchangeDelegate
      *  of the object and releases all held resources.
      *
      */
-    void Shutdown();
+    void Shutdown(bool aAbort = false);
```
Comment (andy31415) [inline src/app/ReadHandler.h:71]: A comment about what 'abort' means would be helpful. As mentioned above, I  believe some constants would be clearer than a bool (kAbortExchange seems obvious to read)

File: src/app/tests/TestReadInteraction.cpp [hunk 4]
```diff
@@ -151,6 +154,16 @@ class MockInteractionModelApp : public chip::app::InteractionModelDelegate
         return err;
     }
 
+    void OnReportData(const chip::app::ReadClient * apReadClient, const chip::app::ClusterInfo & aPath,
+                      chip::TLV::TLVReader * apData, chip::Protocols::InteractionModel::ProtocolCode status) override
+    {
+        mGotAttributeResponse = true;
+    }
+    CHIP_ERROR ReportProcessed(const chip::app::ReadClient * apReadClient) override
```
Comment (andy31415) [inline src/app/tests/TestReadInteraction.cpp:162]: supernit: add one empty line separator between methods for consistency. Above and below.

File: src/app/tests/TestReadInteraction.cpp [hunk 5]
```diff
@@ -161,12 +174,31 @@ class MockInteractionModelApp : public chip::app::InteractionModelDelegate
...
 
 namespace chip {
 namespace app {
+CHIP_ERROR ReadSingleClusterData(ClusterInfo & aClusterInfo, TLV::TLVWriter * apWriter, bool * apDataExists)
+{
+    CHIP_ERROR err   = CHIP_NO_ERROR;
+    uint64_t version = 0;
+    VerifyOrExit(aClusterInfo.mClusterId == kTestClusterId && aClusterInfo.mEndpointId == kTestEndpointId,
+                 err = CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrExit(apWriter != nullptr, /* no op */);
```
Comment (andy31415) [inline src/app/tests/TestReadInteraction.cpp:191]: why is a null apWriter a 'CHIP_NO_ERROR' ?

File: src/app/ReadHandler.cpp (hunk 1 above)
Comment (yunhanw-google) [reply src/app/ReadHandler.cpp:54]: Done, thanks

File: src/app/ReadHandler.cpp (hunk 2 above)
Comment (yunhanw-google) [reply src/app/ReadHandler.cpp:119]: done, thanks

File: src/app/ReadHandler.h (hunk 3 above)
Comment (yunhanw-google) [reply src/app/ReadHandler.h:71]: done, thanks

File: src/app/tests/TestReadInteraction.cpp (hunk 4 above)
Comment (yunhanw-google) [reply src/app/tests/TestReadInteraction.cpp:163]: Done, thanks

File: src/app/tests/TestReadInteraction.cpp (hunk 5 above)
Comment (yunhanw-google) [reply src/app/tests/TestReadInteraction.cpp]: update it with CHIP_ERROR_INVALID_ARGUMENT, thanks

**Review by andy31415 — APPROVED** [review]

**Review by jmartinez-silabs — APPROVED** [review]

**Review by msandstedt — APPROVED** [review]

**Review by jmeg-sfy — APPROVED** [review]

File: src/app/ClusterInfo.h [hunk 6]
```diff
@@ -44,7 +44,14 @@ struct ClusterInfo
...
         }
 
-        if (mFlags.Has(Flags::kFieldIdValid) && other.mFlags.Has(Flags::kFieldIdValid) && (mFieldId == other.mFieldId))
+        if (mFlags.Has(Flags::kFieldIdValid) && !mFlags.Has(Flags::kListIndexValid) && other.mFlags.Has(Flags::kFieldIdValid) &&
+            (mFieldId == other.mFieldId))
+        {
+            return true;
+        }
+
+        if (mFlags.Has(Flags::kFieldIdValid) && mFlags.Has(Flags::kListIndexValid) && other.mFlags.Has(Flags::kListIndexValid) &&
```
Comment (andy31415) [inline src/app/ClusterInfo.h:53]: These conditions are very hard to follow and generally seem to hinge in mFlags optiosn that are repeated. Could you find a way to make them clearer? Even a comment top of the method explaining what fields exist (endpoint, cluster, field, index) and how they relate to each other may help. 

Maybe locally you could use std::optional to combine mFlags and a value (it has  equality operators and an 'ValueOr' functionality if missing is supposed to be a glob value).

Can a ClusterInfo contain no fieldId but a listid? Why do we check of other.ListValid only but for mFlags we check for both fieldid and listid valid?

Given the complexity of the conditionals here, I have a hard time ensuring that the logic is sane. I would rather expect individual checks of 'fieldid match' then 'listindex match' without combining the two (unless combinining is required).

Comment (andy31415) [reply src/app/ClusterInfo.h:53]: Thinking of logic like:

```
Optional<AttributeId> myFieldId = 
    mFlags.has(Flags::kFieldIdValid) 
        ? Optional<AttributeId>::Value(mFieldId);
        : Optional<AttributeId>::Missing();
       
if (myFieldId == Optional<AttributeId>::Value(0xFFFFFFFF)) {
   return true;  /// TODO: above constant should be a constexpr not hardcoded
}

Optional<AttributeId> otherFieldId = ....;

if (myFieldId != otherFieldId) {
  return false;
}

//// setup myListindex and otherListIndex as Optional<ListIndex>

// either both missing or both exactly the same
return (myListIndex == otherListIndex);

```

However I do not know how to solve a potential no field id and valid index... can indexes exist without a field ID or can  indexes exist with a wildcard field id?

I expect this info in a comment explaining the structure and combination possibilities.

**Review by woody-apple — APPROVED** [review]

File: src/app/ClusterInfo.h (hunk 6 above)
Comment (yunhanw-google) [reply src/app/ClusterInfo.h]: Done, thanks

File: src/app/ClusterInfo.h [hunk 7]
```diff
@@ -39,15 +42,44 @@ struct ClusterInfo
...
+        // 1. myListIndex and otherListIndex both missing or both exactly the same, then current is superset of other
+        // 2. myListIndex is missing, no matter if otherListIndex is missing or not, then current is superset of other
+        if (myListIndex == otherListIndex)
         {
+            // either both missing or both exactly the same
             return true;
         }
 
-        if (mFlags.Has(Flags::kFieldIdValid) && other.mFlags.Has(Flags::kFieldIdValid) && (mFieldId == other.mFieldId))
+        if (!myListIndex.HasValue())
```
Comment (andy31415) [inline src/app/ClusterInfo.h:77]: `return !myListIndex.HasValue()` seems to be the last line.

However then why does it return false if myListIndexHaveValue?
I expect 2 identical values (with list index) to return true.

Please add unit tests for the possible combinations of things here.

Comment (andy31415) [reply src/app/ClusterInfo.h:77]: I think the code is  correct: if we get here, the list indices are different. May be worht a comment: difference ok only if mylistindex is missing (missing is a superset of specific)

Comment (yunhanw-google) [reply src/app/ClusterInfo.h:85]: update comment, and the test has been added for these possible combination, thanks
